components/ntp_background_images/browser/ntp_sponsored_image_source_unittest.cc

```diff
@@ -0,0 +1,132 @@
+    feature_list_.InitAndDisableFeature(
+        features::kBraveNTPSuperReferralWallpaper);
+
+    SetUpUrlDataSource();
+
+    SimulateSponsoredComponent();
+  }
+
+  NTPSponsoredImageSource* url_data_source() { return url_data_source_.get(); }
+
+  std::string StartDataRequest(const GURL& url) {
+    CHECK(url_data_source_);
+
+    std::string data;
+    content::WebContents::Getter wc_getter;
+
+    base::RunLoop run_loop;
+    url_data_source_->StartDataRequest(
+        url, wc_getter,
+        base::BindOnce(
+            [](std::string* data, base::OnceClosure quit_closure,
+               scoped_refptr<base::RefCountedMemory> bytes) {
+              if (bytes) {
+                *data = base::as_string_view(*bytes);
+              }
+              std::move(quit_closure).Run();
+            },
+            base::Unretained(&data), run_loop.QuitClosure()));
+    run_loop.Run();
+
+    return data;
+  }
+
+ private:
+  void SetUpUrlDataSource() {
+    NTPBackgroundImagesService::RegisterLocalStatePrefs(
+        pref_service_.registry());
+
+    background_images_service_ = std::make_unique<NTPBackgroundImagesService>(
+        /*component_update_service=*/nullptr, &pref_service_);
+    url_data_source_ = std::make_unique<NTPSponsoredImageSource>(
+        background_images_service_.get());
+  }
+
+  void SimulateSponsoredComponent() {
+    const base::FilePath file_path =
+        test::GetSponsoredImagesComponentPath().AppendASCII("image");
+
+    NTPBackgroundImagesServiceWaiter waiter(*background_images_service_);
+    background_images_service_->OnSponsoredComponentReady(
+        /*is_super_referral=*/false, file_path);
+    waiter.WaitForOnSponsoredImagesDataDidUpdate();
+  }
+
+  content::BrowserTaskEnvironment task_environment_;
+
+  base::test::ScopedFeatureList feature_list_;
+
+  TestingPrefServiceSimple pref_service_;
+
+  std::unique_ptr<NTPBackgroundImagesService> background_images_service_;
+  std::unique_ptr<NTPSponsoredImageSource> url_data_source_;
+};
+
+TEST_F(NTPSponsoredImageSourceTest, StartDataRequest) {
+  const std::string data = StartDataRequest(GURL(
+      R"(chrome://branded-wallpaper/aa0b561e-9eed-4aaa-8999-5627bc6b14fd/background.jpg)"));
+  EXPECT_THAT(data, ::testing::Not(::testing::IsEmpty()));
+}
+
+TEST_F(NTPSponsoredImageSourceTest,
+       DoNotStartDataRequestIfReferencingParentDirectory) {
+  std::string data = StartDataRequest(GURL(
+      R"(chrome://branded-wallpaper/aa0b561e-9eed-4aaa-8999-5627bc6b14fd/campaigns.json)"));
+  EXPECT_THAT(data, ::testing::IsEmpty());
+
+  data = StartDataRequest(GURL(
+      R"(chrome://branded-wallpaper/../aa0b561e-9eed-4aaa-8999-5627bc6b14fd/campaigns.json)"));
+  EXPECT_THAT(data, ::testing::IsEmpty());
+}
+
+TEST_F(NTPSponsoredImageSourceTest,
+       DoNotStartDataRequestIfContentDoesNotExist) {
+  std::string data = StartDataRequest(
+      GURL(R"(chrome://branded-wallpaper/non-existent-file.jpg)"));
+  EXPECT_THAT(data, ::testing::IsEmpty());
+}
```

Comment:
should probably add some tests that requesting files outside of this campaigns directory that DO exist does not succeed

Reply:
We had `DoNotStartDataRequestIfReferencingParentDirectory`, however I have improved the test and renamed to `DoNotStartDataRequestIfContentIsOutsideOfSandbox`. Thanks

Also, see https://github.com/brave/brave-core/blob/master/components/ntp_background_images/browser/ntp_sponsored_source_util_unittest.cc

Reply:
i see, maybe also check that it can't reference things in a different directory at the same level vs just in the parent directory? (since one of the goals here is to prevent cross-campaign resource access)

Reply:
Thanks, added